File: src/main/java/com/example/transactions/domain/Transaction.java
package com.example.transactions.domain;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;

public class Transaction {

    public final Double amount;
    public final Long timestamp;

    public Transaction(Double amount, Long timestamp) {
        this.amount = amount;
        this.timestamp = timestamp;
    }

    public Boolean isNotValid() {
        Instant now = Instant.now();
        return isOlderThan60Seconds(now) || isFuture(now);
    }

    private Boolean isOlderThan60Seconds(Instant now) {
        Instant transactionInstant = Instant.ofEpochMilli(timestamp);
        long diff = transactionInstant.until(now, ChronoUnit.SECONDS);
        return diff > 60;
    }

    private Boolean isFuture(Instant now) {
        Instant transactionInstant = new Date(timestamp).toInstant();
        long diff = transactionInstant.until(now, ChronoUnit.SECONDS);
        return diff < 0;
    }
}

File: domain/src/main/kotlin/com/example/domain/Transaction.kt
package com.example.domain

import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*

class Transaction(val amount: Double, val timestamp: Long) {

    fun isNotValid(): Boolean {
        val now = Instant.now()
        return isOlderThan60Seconds(now) || isFuture(now)
    }

    private fun isOlderThan60Seconds(now: Instant): Boolean {
        val transactionInstant = Instant.ofEpochMilli(timestamp)
        val diff = transactionInstant.until(now, ChronoUnit.SECONDS)
        return diff > 60
    }

    private fun isFuture(now: Instant): Boolean {
        val transactionInstant = Date(timestamp).toInstant()
        val diff = transactionInstant.until(now, ChronoUnit.SECONDS)
        return diff < 0
    }
}
